apps/web/modules/survey/follow-ups/lib/follow-ups.ts

```diff
@@ -0,0 +1,269 @@
+        subject,
+        to: parsedResult.data,
+        replyTo,
+        logoUrl,
+        survey,
+        response,
+        attachResponseData: properties.attachResponseData,
+      });
+
+      return {
+        followUpId,
+        status: "success" as const,
+      };
+    } else {
+      return {
+        followUpId,
+        status: "error",
+        error: `Email address is not valid for followup: ${followUpId}`,
+      };
+    }
+  } else if (Array.isArray(toValueFromResponse)) {
+    const emailAddress = toValueFromResponse[2];
+    if (!emailAddress) {
+      return {
+        followUpId,
+        status: "error",
+        error: `Email address not found in response data for followup: ${followUpId}`,
+      };
+    }
+
+    const parsedResult = z.string().email().safeParse(emailAddress);
+    if (parsedResult.data) {
+      await sendFollowUpEmail({
+        html: body,
+        subject,
+        to: parsedResult.data,
+        replyTo,
+        logoUrl,
+        survey,
+        response,
+        attachResponseData: properties.attachResponseData,
+      });
+
+      return {
+        followUpId,
+        status: "success" as const,
+      };
+    } else {
+      return {
+        followUpId,
+        status: "error",
+        error: `Email address is not valid for followup: ${followUpId}`,
+      };
+    }
+  }
+
+  return {
+    followUpId,
+    status: "error",
+    error: "Something went wrong",
+  };
+};
+
+export const sendSurveyFollowUps = async (
+  organizationId: string,
+  surveyId: string,
+  responseId: string
+): Promise<Result<FollowUpResult[], { code: FollowUpSendError; message: string; meta?: any }>> => {
+  try {
+    await limiter(organizationId);
+  } catch (e) {
+    return err({
+      code: FollowUpSendError.RATE_LIMIT_EXCEEDED,
+      message: "Too many follow‐up requests; please wait a bit and try again.",
+    });
+  }
+
+  validateInputs([surveyId, ZId], [responseId, ZId], [organizationId, ZId]);
+
+  const organization = await getOrganization(organizationId);
+  if (!organization) {
+    return err({
+      code: FollowUpSendError.ORG_NOT_FOUND,
+      message: "Organization not found",
+      meta: {
+        organizationId,
+        surveyId,
+        responseId,
+      },
+    });
+  }
+
+  const surveyFollowUpsPermission = await getSurveyFollowUpsPermission(organization.billing.plan);
+
+  if (!surveyFollowUpsPermission) {
+    return err({
+      code: FollowUpSendError.FOLLOW_UP_NOT_ALLOWED,
+      message: "Survey follow-ups are not allowed for this organization",
+      meta: {
+        organizationId,
+        surveyId,
+        responseId,
+      },
+    });
+  }
+
+  const survey = await getSurvey(surveyId);
+  if (!survey) {
+    return err({
+      code: FollowUpSendError.SURVEY_NOT_FOUND,
+      message: "Survey not found",
+      meta: {
+        organizationId,
+        surveyId,
+        responseId,
+      },
+    });
+  }
+
+  const response = await getResponse(responseId);
```

Comment:
I think it would be more efficient to have one good query call that gets the response plus all other information (survey, org) from the db.